scoring/score.py

```diff
@@ -13,21 +25,182 @@ def __init__(self, message: str, *, code: str) -> None:
+    of names.
+
+    >>> join_text(["foo", "bar", "baz"], "and")
+    "foo, bar and baz"
+
+    >>> join_text(["foo", "bar"], "or")
+    "foo or bar"
+
+    >>> join_text(["foo"], "or")
+    "foo"
+    """
+    strings = tuple(strings)
+
+    try:
+        *strings, right = strings
+    except ValueError:
+        return ""
+
+    if not strings:
+        return right
+
+    left = ", ".join(strings[:-1])
+    return f"{left} {separator} {strings[-1]}"
+
+
+def join_and(strings: Iterable[str]) -> str:
+    return join_text(strings, "and")
+
+
+def join_or(strings: Iterable[str]) -> str:
+    return join_text(strings, "or")
+
+
 class Scorer:
     def __init__(self, teams_data, arena_data):
         self._teams_data = teams_data
-        self._arena_data = arena_data
+        self._districts = arena_data['other']['districts']
+
+        for district in self._districts.values():
+            district['pallets'] = collections.Counter(district['pallets'])
+            district['highest'] = district['highest'].replace(' ', '')
+
+    def score_district_for_zone(self, name: str, district: RawDistrict, zone: int) -> int:
+        colour = ZONE_COLOURS[zone]
+
+        num_tokens = district['pallets'][colour]
+        score = num_tokens * DISTRICT_SCORE_MAP[name]
+
+        if colour in district['highest']:
+            # Points are doubled for the team owning the highest pallet
+            score *= 2
+
+        return score
 
     def calculate_scores(self):
         scores = {}
 
         for tla, info in self._teams_data.items():
-            scores[tla] = 0
+            district_score = sum(
+                self.score_district_for_zone(name, district, info['zone'])
+                for name, district in self._districts.items()
+            )
+            movement_score = 1 if info.get('left_starting_zone') else 0
+            scores[tla] = district_score + movement_score
 
         return scores
 
     def validate(self, other_data):
-        warnings.warn("Scoresheet validation not implemented")
+        # Check that the right districts are specified.
+        if self._districts.keys() != DISTRICT_SCORE_MAP.keys():
+            missing = DISTRICT_SCORE_MAP.keys() - self._districts.keys()
+            extra = self._districts.keys() - DISTRICT_SCORE_MAP.keys()
+            detail = "Wrong districts specified."
+            if missing:
+                detail += f" Missing: {join_and(missing)}."
+            if extra:
+                detail += f" Extra: {join_and(repr(x) for x in extra)}."
+            raise InvalidScoresheetException(
+                detail,
+                code='invalid_districts',
+            )
+
+        # Check that the "highest" pallet is a single, valid colour entry.
+        bad_highest = {}
+        for name, district in self._districts.items():
+            highest = district['highest']
+            if highest and highest not in ZONE_COLOURS:
```

Comment:
[nit] I think this is also valid?
```suggestion
            if highest not in ZONE_COLOURS:
```

Reply:
I don't think we want to alert about cases where `highest` is empty string (`''`) -- implying that there was no highest recorded? The validity of not having a highest pallet is checked elsewhere.

Reply:
oop you're right, I think I misread it, never mind!